### DISCUSSION TITLE
How can I get some resources with the event type?

### DISCUSSION THREAD

**Discussion by terassyi** [kube-rs/kube#1312] [Q&A] [answered]
Hi, I'm trying to watch deleting Endpointslice resources.
And my code is like this.
```rust
    let endpointslices = Api::<EndpointSlice>::all(client);
    tracing::info!(name=eps.name_any(),obj=?eps,"Reconcile Endpointslice");

    finalizer(
        &endpointslices,
        ENDPOINTSLICE_FINALIZER,
        eps,
        |event| async {
            match event {
                Event::Apply(eps) => reconcile(&eps, ctx.clone()).await,
                Event::Cleanup(eps) => cleanup(&eps, ctx.clone()).await,
            }
        },
    )
    .await
    .unwrap();
```
I tried to get deleting resources with the finalizer function.
However, I got the error logs below and cannot add the finalizer to Endpointslice resources.

```
2023-10-12T14:42:48.275224Z  WARN reconciling object{object.ref=EndpointSlice.v1.discovery.k8s.io/app-svc-local-cmffz.test object.reason=object updated}: sartd::kubernetes::context: reconcile failed: FinalizerError(AddFinalizer(Api(ErrorResponse { status: "Failure", message: "the server could not find the requested resource", reason: "NotFound", code: 404 })))
```

I'm not sure a way to watch resources with the event type, such as applying, updating, and deleting.

How can I watch and get events and resources?

Thank you!

**Comment by clux** [comment, marked as answer]
Hey. The `finalizer` fn is to help abstract the management of Kubernetes finalizers, not to give precise watch events.
There are some docs on https://kube.rs/controllers/gc/#finalizers which talks about the `secret_syncer` example.

I think your error is passing an `Api:all` to `finalizer`, but you have a namespaced resource. The `Api::all` can only be used to view events across namespaces, not to patch namespaced resources (see https://github.com/kube-rs/kube/issues/1030). Note how `secret_syncer` is creating an `Api::namespaced` for the namespace of the object.

**Reply by terassyi** [reply]
I appreciate your help!

I fully misunderstood the usage of `Api::all()` and `Api::namespaced().`
And I could reconcile Endpointslices.

Thank you.
